generate/lora.py

```diff
@@ -0,0 +1,154 @@
+import warnings
+from functools import partial
+from pathlib import Path
+from typing import Literal
+
+import lightning as L
+import torch
+from lightning.fabric.strategies import FSDPStrategy
+from torch.distributed.fsdp.wrap import transformer_auto_wrap_policy
+
+# support running without installing as a package
+wd = Path(__file__).parent.parent.resolve()
+sys.path.append(str(wd))
+
+from generate.base import generate
+from lit_parrot import Tokenizer
+from lit_parrot.lora import lora
+from lit_parrot.model import Parrot, Config
+from lit_parrot.utils import EmptyInitOnDevice, lazy_load, check_valid_checkpoint_dir
+from scripts.prepare_alpaca import generate_prompt
+
+
+lora_r = 8
+lora_alpha = 16
+lora_dropout = 0.05
+
+
+def main(
+    prompt: str = "What food do lamas eat?",
+    input: str = "",
+    lora_path: Path = Path("out/lora/alpaca/lit-model-lora-finetuned.pth"),
+    checkpoint_dir: Path = Path(f"checkpoints/stabilityai/stablelm-base-alpha-3b"),
+    quantize: Literal["llm.int8", "gptq.int4"] = None,
+    max_new_tokens: int = 100,
+    top_k: int = 200,
+    temperature: float = 0.8,
+    strategy: str = "auto",
+    devices: int = 1,
+    precision: str = "bf16-true",
+) -> None:
+    """Generates a response based on a given instruction and an optional input.
+    This script will only work with checkpoints from the instruction-tuned Parrot-Adapter model.
+    See `finetune/adapter.py`.
+
+    Args:
+        prompt: The prompt/instruction (Alpaca style).
+        adapter_path: Path to the checkpoint with trained adapter weights, which are the output of
+            `finetune/adapter.py`.
+        checkpoint_dir: The path to the checkpoint folder with pretrained Parrot weights.
+        input: Optional input (Alpaca style).
+        quantize: Whether to quantize the model and using which method:
+            ``"llm.int8"``: LLM.int8() mode,
+            ``"gptq.int4"``: GPTQ 4-bit mode.
+        max_new_tokens: The number of generation steps to take.
+        top_k: The number of top most probable tokens to consider in the sampling process.
+        temperature: A value controlling the randomness of the sampling process. Higher values result in more random
+            samples.
+        strategy: Indicates the Fabric strategy setting to use.
+        devices: How many devices to use.
+        precision: Indicates the Fabric precision setting to use.
+    """
+    if strategy == "fsdp":
+        auto_wrap_policy = partial(transformer_auto_wrap_policy, transformer_layer_cls={Block})
+        strategy = FSDPStrategy(auto_wrap_policy=auto_wrap_policy, cpu_offload=False)
+    fabric = L.Fabric(devices=devices, precision=precision, strategy=strategy)
+    fabric.launch()
+
+    check_valid_checkpoint_dir(checkpoint_dir)
+
+    with open(checkpoint_dir / "lit_config.json") as fp:
+        config = Config(**json.load(fp))
+
+    if quantize is not None:
+        raise NotImplementedError("Quantization in LoRA is not supported yet")
```

Comment:
In this case, I would remove the argument entirely

Reply:
It might actually work, it's unclear. Can't test it at the moment due to import issues with bitsandbytes. Probably related to https://github.com/openai/triton/issues/1735